testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/AppInitiatedActionCancelTest.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.testsuite.actions;
+
+import org.jboss.arquillian.graphene.page.Page;
+import org.junit.Test;
+import org.keycloak.models.UserModel;
+import org.keycloak.representations.idm.RealmRepresentation;
+import org.keycloak.testsuite.pages.LoginUpdateProfileEditUsernameAllowedPage;
+
+/**
+ * Test makes sure that sending a cancel signal does not remove a non-AIA
+ * required action
+ *
+ * @author Stan Silvert
+ */
+public class AppInitiatedActionCancelTest extends AbstractAppInitiatedActionTest {
+    
+    @Page
+    protected LoginUpdateProfileEditUsernameAllowedPage updateProfilePage;
+    
+    public AppInitiatedActionCancelTest() {
+        super("update_profile");
+    }
+    
+    @Override
+    public void configureTestRealm(RealmRepresentation testRealm) {
+        ActionUtil.addRequiredActionForUser(testRealm, "test-user@localhost", UserModel.RequiredAction.UPDATE_PROFILE.name());
+    }
+    
+    @Test
+    // Verify that sending a "cancel" does not remove the required action.
+    public void cancelUpdateProfile() {
+        doAIA();
+        loginPage.login("test-user@localhost", "password");
+        updateProfilePage.assertCurrent();
+        updateProfilePage.cancel();
+        assertRedirectSuccess();
```

Comment:
This should check that an error response is sent from Keycloak (https://openid.net/specs/openid-connect-core-1_0.html#AuthError) and that the error code/description indicates to the app it was the AIA that was cancelled and not the login.

Reply:
In this case when the update profile is required it should not show the cancel button. The cancel button should only be displayed/processed in the case an action is not required, but just initiated by AIA.

Reply:
I'll fix the test.

I need to look into your second concern.  Not sure if that's broken.

Reply:
I implemented this.  It only works for OIDC, not SAML.

Also, I added a silent mode.  This is needed because the OIDC error message screws up the URL.  For the account console, this looks really bad.  So I give the option to not add the message in case of cancel.